sources/packages/backend/apps/workers/src/services/student-assessment/student-assessment.service.ts

```diff
@@ -49,15 +53,41 @@ export class StudentAssessmentService extends RecordDataModelService<StudentAsse
           ASSESSMENT_NOT_FOUND,
         );
       }
+      // Check if the assessment is in one of the initial statuses. Any status different than "Submitted" and "Queued" should
+      // cancel the workflow because there is no point executing it.
+      // A workflow in these statuses can be associated and updated to "In progress" nicely. The expectation is that it would
+      // be primarily "Queued" but there is no harm in allowing "Submitted" also.
+      // The main intention is to prevent the workflow association from happening in any other status, for instance,
+      // for the case when workers are down and a workflow is triggered, the assessment can have its status updated
+      // (e.g. be canceled) before the workers are up and running again, which can potentially make the workflow execution no longer needed.
+      const initialStatuses = [
+        StudentAssessmentStatus.Submitted,
+        StudentAssessmentStatus.Queued,
+      ];
+      if (!initialStatuses.includes(assessment.studentAssessmentStatus)) {
+        throw new CustomNamedError(
+          `The assessment is not in any initial status that would allow the workflow association. Expected statuses: ${initialStatuses.join(
+            ", ",
+          )}.`,
+          INVALID_OPERATION_IN_THE_CURRENT_STATUS,
+        );
+      }
       if (!assessment.assessmentWorkflowId) {
         // Assessment is available to be associated with the workflow instance id.
-        await assessmentRepo.update(assessmentId, { assessmentWorkflowId });
+        const now = new Date();
+        await assessmentRepo.update(assessmentId, {
+          assessmentWorkflowId,
+          studentAssessmentStatus: StudentAssessmentStatus.InProgress,
+          studentAssessmentStatusUpdatedOn: now,
+          modifier: auditUser,
+          updatedAt: now,
+        });
         return;
       }
       if (assessment.assessmentWorkflowId !== assessmentWorkflowId) {
         throw new CustomNamedError(
           `The assessment is already associated with another workflow instance. Current associated instance id ${assessment.assessmentWorkflowId}.`,
-          ASSESSMENT_INVALID_OPERATION_IN_THE_CURRENT_STATE,
+          ASSESSMENT_ALREADY_ASSOCIATED_WITH_DIFFERENT_WORKFLOW,
         );
       }
       // The workflow was already associated with the workflow, no need to update it again.
```

Comment:
I was wondering if this part could refactored to put all the validations before the association like in: 
![image](https://github.com/bcgov/SIMS/assets/78114138/83fe764a-c5b2-4825-930d-015b10c82462)

Reply:
I guess so, I will take a look 😉

Reply:
Method refactored, please take a look 😉